sql-plugin/src/main/scala/com/nvidia/spark/rapids/AstUtil.scala

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeSeq, Expression, NamedExpression}
+import org.apache.spark.sql.rapids.catalyst.expressions.{GpuEquivalentExpressions, GpuExpressionEquals}
+
+
+object AstUtil {
+
+  /**
+   * Check whether it can be split into non-ast sub-expression if needed
+   *
+   * @return true when: 1) If all ast-able in expr; 2) all non-ast-able tree nodes don't contain
+   *         attributes from both join sides. In such case, it's not able
+   *         to push down into single child.
+   */
+  def canExtractNonAstConditionIfNeed(expr: BaseExprMeta[_], left: Seq[Attribute],
```

Comment:
It would really be nice to provide a way with tagging to show the user why an expression does not work with this. I was trying to play around with it and was doing.

```
spark.time(left.join(right, length(col("leftstr")) + length(col("rightstr")) >= left("id") + right("id"), joinType="leftouter").count())
```

But it would not put the expression on the GPU because 

```
            !Expression <Cast> cast((length(leftstr#301) + length(rightstr#296)) as bigint) cannot run on GPU because AST is required and this expression does not support AST
              @Expression <Add> (length(leftstr#301) + length(rightstr#296)) could run on GPU
                !Expression <Length> length(leftstr#301) cannot run on GPU because AST is required and this expression does not support AST
                  !Expression <AttributeReference> leftstr#301 cannot run on GPU because AST is required and expression AttributeReference leftstr#301 produces an unsupported type StringType
                !Expression <Length> length(rightstr#296) cannot run on GPU because AST is required and this expression does not support AST
                  !Expression <AttributeReference> rightstr#296 cannot run on GPU because AST is required and expression AttributeReference rightstr#296 produces an unsupported type StringType
```

It took a bit to be able to see that the cast that was inserted after the add could not be on the GPU so the entire expression could not be translated to the GPU, and that was after I had read through the code.  I am fine if this is follow on work.